File: extended_gcd.py
def extended_gcd(a: int, b: int, positive_x: bool)->tuple[int,int,int]:

    original_a = a
    original_b = b
    
    # We want to find x and y such that ax + by = gcd(a,b)
    # if b > a then the result output will swap and so will the signs of the output.   
    # previous x and y
    x0, y0 = 1,0 #represents a when x = 1 and y = 0
    # current x and y
    x1, y1 = 0,1 #represents b when x = 0 and y = 1

    while b !=0: #while the remainder isn't 0......
        # This part is same as ordinary GCD algorithm. 
        # quotient = whole number floor result of a divided by b
        quotient = a//b
        # set a to b and at same time set remainder b = a mod b
        a, b = b, a % b
        
        # This part for finding the coefficients by setting current values based on previous values.
        # We are keeping track of expressing each remainder in terms of current a and b.
        # This is why we are referring back to the previous values.
        # current x0 becomes last x1 and current x1 becomes last x0 - current quotient x last x1
        # This is because the new remainder is previous a - quotient x previous b.
        x0, x1 = x1, x0 - quotient * x1
        # current y0 becomes last y1 and current y1 becomes last y0 - current quotient x last y1
        y0, y1 = y1, y0 - quotient * y1

    # If we want positive x but we have negative x as a result, we can use:
    # x = x0 + k*(b/gcd)
    # y = y0 + k*(a/gcd)
    # k will represent a shift of multiples of b/gcd (or would be a/gcd if we wanted positive y)
    # needed if we are wanting a*x = 1 (mod original_b)
    if (positive_x and x0 < 0):
       # Shift of k calculates how much needs to be added in one go where original_b//a is size of step needed
       # use (original_b//a) - 1 to make sure we get shift rounded up so x0 becomes positive.
       shift_k = (abs(x0) + (original_b // a) - 1) // (original_b // a)  # smallest integer needed to make x positive
       x0 += shift_k*(original_b//a) #a = gcd
       y0 -= shift_k*(original_a//a)
            
            
    return a,x0,y0

File: extended_gcd.rs
fn extended_gcd(mut a: i32, mut b: i32, positive_x: bool) -> (i32, i32, i32) {
    let original_a = a;
    let original_b = b;
    let mut x0 = 0;
    let mut y0 = 1;
    let mut x1 = 1;
    let mut y1 = 0;

    while b != 0 {
        let quotient = a / b;
        let (new_a, new_b) = (b, a % b);
        a = new_a;
        b = new_b;

        let temp_x0 = x0;
        x0 = x1;
        x1 = temp_x0 - quotient * x1;

        let temp_y0 = y0;
        y0 = y1;
        y1 = temp_y0 - quotient * y1;
    }

    if positive_x && x0 < 0 {
        let shift_k = (x0.abs() + (original_b / a) - 1) / (original_b / a);
        x0 += shift_k * (original_b / a);
        y0 -= shift_k * (original_a / a);
    }

    (a, x0, y0)
}